File: demo-apps/iOS/SimpleObjC/App/ViewControllers/ConnectionViewController.m
//
//  ConnectionViewController.m
//  SimpleChatObjC
//
//  Connection screen for SimpleChatObjC
//

#import "ConnectionViewController.h"
#import "ConversationListViewController.h"
#import <FinClipChatKit/FinClipChatKit-Swift.h>

@interface ConnectionViewController () <CKTChatKitCoordinatorDelegate>

@property (nonatomic, strong) CKTChatKitCoordinator *coordinator;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *serverHintLabel;
@property (nonatomic, strong) UIButton *connectButton;
@property (nonatomic, strong) UILabel *statusLabel;

@end

@implementation ConnectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupUI];
}

- (void)setupUI {
    self.view.backgroundColor = UIColor.systemBackgroundColor;
    self.title = @"Connect";
    
    // Title label
    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.text = @"SimpleChatObjC";
    self.titleLabel.font = [UIFont systemFontOfSize:28 weight:UIFontWeightBold];
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.titleLabel];
    
    // Server hint label
    self.serverHintLabel = [[UILabel alloc] init];
    self.serverHintLabel.text = @"Connecting to: http://127.0.0.1:3000/agent";
    self.serverHintLabel.font = [UIFont systemFontOfSize:14];
    self.serverHintLabel.textColor = UIColor.secondaryLabelColor;
    self.serverHintLabel.textAlignment = NSTextAlignmentCenter;
    self.serverHintLabel.numberOfLines = 0;
    self.serverHintLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.serverHintLabel];
    
    // Connect button
    UIButtonConfiguration *buttonConfig = [UIButtonConfiguration filledButtonConfiguration];
    buttonConfig.title = @"Connect";
    buttonConfig.cornerStyle = UIButtonConfigurationCornerStyleMedium;
    self.connectButton = [UIButton buttonWithConfiguration:buttonConfig primaryAction:nil];
    [self.connectButton addTarget:self 
                           action:@selector(connectTapped)
                 forControlEvents:UIControlEventTouchUpInside];
    self.connectButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.connectButton];
    
    // Status label
    self.statusLabel = [[UILabel alloc] init];
    self.statusLabel.text = @"Ready to connect";
    self.statusLabel.font = [UIFont systemFontOfSize:14];
    self.statusLabel.textColor = UIColor.secondaryLabelColor;
    self.statusLabel.textAlignment = NSTextAlignmentCenter;
    self.statusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.statusLabel];
    
    // Layout constraints
    [NSLayoutConstraint activateConstraints:@[
        [self.titleLabel.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:60],
        [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:20],
        [self.titleLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20],
        
        [self.serverHintLabel.topAnchor constraintEqualToAnchor:self.titleLabel.bottomAnchor constant:40],
        [self.serverHintLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:40],
        [self.serverHintLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-40],
        
        [self.connectButton.topAnchor constraintEqualToAnchor:self.serverHintLabel.bottomAnchor constant:30],
        [self.connectButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.connectButton.widthAnchor constraintEqualToConstant:200],
        [self.connectButton.heightAnchor constraintEqualToConstant:50],
        
        [self.statusLabel.topAnchor constraintEqualToAnchor:self.connectButton.bottomAnchor constant:20],
        [self.statusLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:20],
        [self.statusLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20]
    ]];
}

- (void)connectTapped {
    self.connectButton.enabled = NO;
    self.statusLabel.text = @"Connecting...";
    
    // Initialize SDK coordinator with default configuration
    NSURL *serverURL = [NSURL URLWithString:@"http://127.0.0.1:3000/agent"];
    CKTCoordinatorConfig *config = [[CKTCoordinatorConfig alloc] initWithServerURL:serverURL
                                                                             userId:@"demo-user"
                                                                           deviceId:nil];
    config.storageMode = CKTStorageModePersistent;
    
    self.coordinator = [[CKTChatKitCoordinator alloc] initWithConfig:config];
    self.coordinator.delegate = self;
    
    // Navigate to conversation list
    ConversationListViewController *listVC = [[ConversationListViewController alloc] initWithSDKCoordinator:self.coordinator];
    listVC.autoCreateConversation = YES;
    [self.navigationController pushViewController:listVC animated:YES];
    
    self.connectButton.enabled = YES;
}

#pragma mark - CKTChatKitCoordinatorDelegate

- (void)chatKitCoordinator:(CKTChatKitCoordinator *)coordinator didChangeState:(NSString *)state {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.statusLabel.text = state;
    });
}

- (void)chatKitCoordinator:(CKTChatKitCoordinator *)coordinator didEncounterError:(NSError *)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Connection Error"
                                                                       message:error.localizedDescription
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
    });
}

@end

File: demo-apps/iOS/Simple/App/ViewControllers/ConnectionViewController.swift
import UIKit
import Combine
import FinClipChatKit

final class ConnectionViewController: UIViewController {
  private var coordinator: ChatKitCoordinator?
  private var cancellables = Set<AnyCancellable>()

  private lazy var titleLabel: UILabel = {
    let label = UILabel()
    label.text = "Simple"
    label.font = .systemFont(ofSize: 28, weight: .bold)
    label.textAlignment = .center
    label.translatesAutoresizingMaskIntoConstraints = false
    return label
  }()
  
  private lazy var serverHintLabel: UILabel = {
    let label = UILabel()
    label.text = "Connecting to: http://127.0.0.1:3000/agent"
    label.font = .systemFont(ofSize: 14)
    label.textColor = .secondaryLabel
    label.textAlignment = .center
    label.numberOfLines = 0
    label.translatesAutoresizingMaskIntoConstraints = false
    return label
  }()

  private lazy var connectButton: UIButton = {
    var config = UIButton.Configuration.filled()
    config.title = "Connect"
    config.cornerStyle = .medium
    let button = UIButton(configuration: config)
    button.addTarget(self, action: #selector(connectTapped), for: .touchUpInside)
    button.translatesAutoresizingMaskIntoConstraints = false
    return button
  }()

  private lazy var statusLabel: UILabel = {
    let label = UILabel()
    label.text = "Ready to connect"
    label.font = .systemFont(ofSize: 14)
    label.textColor = .secondaryLabel
    label.textAlignment = .center
    label.translatesAutoresizingMaskIntoConstraints = false
    return label
  }()

  override func viewDidLoad() {
    super.viewDidLoad()
    setupUI()
    setupCoordinator()
  }

  private func setupUI() {
    view.backgroundColor = .systemBackground
    title = "Connect"

    view.addSubview(titleLabel)
    view.addSubview(serverHintLabel)
    view.addSubview(connectButton)
    view.addSubview(statusLabel)

    NSLayoutConstraint.activate([
      titleLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 60),
      titleLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
      titleLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),

      serverHintLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 40),
      serverHintLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 40),
      serverHintLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -40),

      connectButton.topAnchor.constraint(equalTo: serverHintLabel.bottomAnchor, constant: 30),
      connectButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
      connectButton.widthAnchor.constraint(equalToConstant: 200),
      connectButton.heightAnchor.constraint(equalToConstant: 50),

      statusLabel.topAnchor.constraint(equalTo: connectButton.bottomAnchor, constant: 20),
      statusLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
      statusLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20)
    ])
  }

  private func setupCoordinator() {
    // Coordinator will be created when user connects
    // Network state monitoring will be set up after coordinator is created
  }

  @objc private func connectTapped() {
    connectButton.isEnabled = false
    statusLabel.text = "Connecting..."

    // Create coordinator
    let config = NeuronKitConfig.default(serverURL: AppConfig.defaultServerURL)
        .withUserId(AppConfig.defaultUserId)
    let coordinator = ChatKitCoordinator(config: config)
    self.coordinator = coordinator
    
    // Monitor network state
    coordinator.runtime.networkStatePublisher()
      .receive(on: DispatchQueue.main)
      .sink { [weak self] state in
        guard let self = self else { return }
        switch state {
        case .connected:
          self.statusLabel.text = "Connected"
        case .connecting:
          self.statusLabel.text = "Connecting..."
        case .reconnecting:
          self.statusLabel.text = "Reconnecting..."
        case .disconnected:
          self.statusLabel.text = "Disconnected"
        case .error(let error):
          self.statusLabel.text = "Error"
          self.showAlert("Connection Error", message: error.message)
        @unknown default:
          self.statusLabel.text = "Unknown state"
        }
      }
      .store(in: &cancellables)

    let listVC = ConversationListViewController(coordinator: coordinator)
    listVC.autoCreateConversation = true
    navigationController?.pushViewController(listVC, animated: true)

    connectButton.isEnabled = true
  }

  private func showAlert(_ title: String, message: String) {
    let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
    alert.addAction(UIAlertAction(title: "OK", style: .default))
    present(alert, animated: true)
  }
}
